### PR TITLE
[compiler-v2] Introduce language version

### PR DESCRIPTION
**Pull request by wrwg**
## Description

This introduces the concept of a language version and pipes it through the various layers of the stack (ugh! we need to refactor this mountains of different Move config options), from aptos CLI down to compiler v2, prover, and transactional test framework.

- The `LanguageVersion` resides in the new file in the Move model, `metadata.rs`. The `CompilerVersion` has been also moved from the package system into there.  Compiler and language version support formatting and parsing. They also support their default being populated from env vars `MOVE_COMPILER_V2` and `MOVE_LANGUAGE_V2`, respectively.
- There is also a new data structure `CompilationMetadata` which we can use to attach to the file format metadata section in subsequent PRs.
- Both compiler and language version have a qualifier whether a particular version is 'stable'. The value of this is populated in `CompilationMetadata` so we can perform run time checks and disallow an unstable version on production chains (subsequent PRs).
- The aptos cli will print out a warning if an unstable compiler or language version is used.
- The language version is attached to the GlobalEnv. The expression builder will refuse to build language constructs not supported by a given language version.
- File format generator will not generate access specifiers if language v2 is not selected or if a special new experiment GEN_ACCESS_SPECIFIERS is on. This mitigates #12623 for now so we can lower priority but does not close it.
- The acquires checks have been adapted to use the language version in the exp_builder instead of the previous flag.
- A new test directory `checking-lang-v1` has been created to contain tests checking for disallowed v2 language features; some acquires related tests have been move there as well.

## Type of Change
- [x] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Existing tests and new tests

## Key Areas to Review

Baselines files as code changes to key sources like `metadata.rs`.

### CODE REVIEW

File: third_party/move/move-compiler-v2/tests/checking-lang-v1/conditional_global_operations.exp [hunk 1]
```diff
@@ -1,25 +1,25 @@
 
 Diagnostics:
-error: access specifier not enabled. Only plain `acquires <ResourceName>` is enabled.
-   ┌─ tests/acquires-checker/conditional_global_operations.move:14:49
+error: access specifiers not supported in language versions before `2.0-unstable`
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/tests/checking-lang-v1/conditional_global_operations.exp:3]: As given, this error message is bound to confuse the reader. It would seem that `acquires` is indeed supported in older versions of the language, it is just that instantiated types are not allowed. Maybe there is already an issue to tackle this, if not, it would be good to file one.

File: third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs [hunk 2]
```diff
@@ -418,11 +429,34 @@ impl ModuleGenerator {
...
+                    .collect()
+            })
+        } else {
+            // Report an error if we cannot drop the access specifiers.
+            // TODO(#12623): remove this once the bug is fixed
+            if fun_env
+                .get_access_specifiers()
+                .map(|v| {
+                    v.iter().any(|s| {
+                        s.kind != AccessKind::Acquires
```
Comment (vineethk) [inline third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs:445]: This looks similar to `AccessSpecifier::is_old_style_acquires()`. Use that instead?

File: third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs [hunk 3]
```diff
@@ -448,8 +449,14 @@ static MOVE_STDLIB_COMPILED: Lazy<Vec<CompiledModule>> = Lazy::new(|| {
 #[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq)]
 pub enum TestRunConfig {
     CompilerV1,
-    CompilerV2 { v2_experiments: Vec<(String, bool)> },
-    ComparisonV1V2 { v2_experiments: Vec<(String, bool)> },
+    CompilerV2 {
```
Comment (vineethk) [inline third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs:452]: I think this could be restructured to explicitly involve compiler version enums (and allow for comparisons between various combinations), which allow for future compiler versions as well.

But maybe this is a follow up task and not for this PR.

**Review by vineethk — COMMENTED** [review]
Looking good so far.

**Comment by rahxephon89** [comment]
By running e2e-move-test with V7 bytecode, I can confirm that #12623 is mitigated by this PR

File: third_party/move/move-model/src/metadata.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+};
+
+const UNSTABLE_MARKER: &str = "-unstable";
+
+/// Metadata about a compilation result. To maintain serialization
+/// compatibility, this uses a free-form string to represent compiler version
+/// and language version.
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct CompilationMetadata {
+    /// The version of the compiler, as a string.
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/metadata.rs:21]: The syntax is really determined by the types declared later, but this need to be stable in upwards and downwards compatible deserialization scenarios, so strings are used here. I clarified in the comment.

File: third_party/move/move-model/src/metadata.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    pub compiler_version: String,
+    /// The version of the language, as a string.
+    pub language_version: String,
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/metadata.rs:24]: See above

File: third_party/move/move-model/src/metadata.rs [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    /// The legacy v1 Move compiler
+    V1,
+    /// The v2 compiler, starting with 2.0. Each new released version of the compiler
+    /// should get an enum entry here.
+    V2_0,
+}
+
+impl Default for CompilerVersion {
+    /// We allow the default to be set via an environment variable.
+    fn default() -> Self {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/metadata.rs:66]: Maybe later, when we have more than 2 versions. Can be easily extended as needed.

File: third_party/move/move-model/src/metadata.rs [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+impl FromStr for CompilerVersion {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Strip unstable marker as it is not relevant for parsing.
+        let s1 = s.replace(UNSTABLE_MARKER, "");
+        match s1.as_str() {
+            // For legacy reasons, also support v1 and v2
+            "1" | "v1" => Ok(Self::V1),
+            "2" | "v2" | "2.0" => Ok(Self::V2_0),
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/metadata.rs:85]: - Its always only two digits. But yes, 2 is 2.latest. We can flesh this out as we move.
- The v1/v2 is really only for cli api compatibility. We would need to find where its used (some github actions perhaps) but this can be done in another PR.

File: third_party/move/move-model/src/metadata.rs [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        }
+    }
+}
+
+/// Represents a language version.
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, Eq, PartialEq, PartialOrd, Ord)]
+pub enum LanguageVersion {
+    /// The version of Move at around the genesis of the Aptos network end
+    /// of '22. This is the original Diem Move plus the extension of inline
+    /// functions with lambda parameters.
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/metadata.rs:139]: Done

File: third_party/move/move-model/src/metadata.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/metadata.rs:85]: Ok, I clarified the versioning scheme for both compiler and language version.

File: third_party/move/move-model/src/model.rs [hunk 9]
```diff
@@ -602,6 +608,7 @@ impl GlobalEnv {
         let unknown_move_ir_loc = MoveIrLoc::new(FileHash::new("<unknown>"), 0, 0);
         let internal_loc = fake_loc("<internal>");
         GlobalEnv {
+            language_version: LanguageVersion::V1,
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/model.rs:611]: Done

File: third_party/move/move-compiler-v2/tests/checking-lang-v1/conditional_global_operations.exp (hunk 1 above)
Comment (wrwg) [reply third_party/move/move-compiler-v2/tests/checking-lang-v1/conditional_global_operations.exp:3]: Improved the messages

File: third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs (hunk 2 above)
Comment (wrwg) [reply third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs:445]: Unfortunately, that one is on the file_format version of the thing, whereas here we look at the AST with location info. This will go away, so also no point to factor it out.

File: third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs (hunk 3 above)
Comment (wrwg) [reply third_party/move/testing-infra/transactional-test-runner/src/vm_test_harness.rs:452]: Good idea, but wasn't really the objective of this PR.

**Comment by wrwg** [comment]
PTAL.

When reviewing, perhaps pay not too much attention to the proposed `CompilationMetadata` type. This is for now connected to nothing and can be refined when we connect this to the file format.

The important and by far larger part of the PR is the new version types and plumbing of the `LanguageVersion` through the stack, as well as changing the compiler to support the language version.

**Review by rahxephon89 — APPROVED** [review]
LGTM, thanks!

File: third_party/move/move-model/src/metadata.rs [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+    /// Returns true of the compilation was created as unstable.
+    pub fn created_as_unstable(&self) -> bool {
+        self.unstable
+    }
+}
+
+/// Represents a compiler version
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, Eq, PartialEq, PartialOrd, Ord)]
+pub enum CompilerVersion {
```
Comment (ziaptos) [inline third_party/move/move-model/src/metadata.rs:71]: Should we do the Semantic versioning here? Maybe ok to have every shingle Major/Minor/Patch - so:
V1_0_0
V2_0_0_Alpha_1 ... etc

Comment (ziaptos) [reply third_party/move/move-model/src/metadata.rs:71]: Alternatively we can call this enum CompilerVersionMajor

File: third_party/move/move-model/src/metadata.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ziaptos) [reply third_party/move/move-model/src/metadata.rs:102]: +1

Let's just use the standard 3-part Semantic versioning.

File: third_party/move/move-model/src/metadata.rs [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    /// should not be allowed on production networks.
+    pub fn unstable(self) -> bool {
+        match self {
+            LanguageVersion::V1 => false,
+            LanguageVersion::V2_0 => true,
+        }
+    }
+}
+
+impl Display for LanguageVersion {
```
Comment (ziaptos) [inline third_party/move/move-model/src/metadata.rs:217]: nit: can we just have some static tables with both display string and boolean for stable/unstable? And avoid all this (and above) logic?

File: third_party/move/move-model/src/model.rs [hunk 12]
```diff
@@ -637,11 +644,21 @@ impl GlobalEnv {
         self.generated_by_v2 = yes
     }
 
-    /// Returns if this is generated by v2.
+    /// Returns true if this is generated by v2.
     pub fn generated_by_v2(&self) -> bool {
```
Comment (ziaptos) [inline third_party/move/move-model/src/model.rs:645]: nit: rename to `is_generated_by_v2`, and remove the comment.

File: third_party/move/move-model/src/model.rs [hunk 13]
```diff
@@ -637,11 +644,21 @@ impl GlobalEnv {
...
         self.generated_by_v2
     }
 
+    /// Sets the language version.
+    pub fn set_language_version(&mut self, version: LanguageVersion) {
+        self.language_version = version
+    }
+
+    /// Returns the language version
```
Comment (ziaptos) [inline third_party/move/move-model/src/model.rs:654]: nit: same - no need to comment functions with `self commenting` names

File: third_party/move/move-model/src/options.rs [hunk 14]
```diff
@@ -11,6 +12,9 @@ pub struct ModelBuilderOptions {
     /// part of the model.
     pub compile_via_model: bool,
 
+    /// The language version to use.
+    pub language_version: LanguageVersion,
```
Comment (ziaptos) [inline third_party/move/move-model/src/options.rs:16]: nit: `pub language_version_to_use: LanguageVersion`; and no comment

File: third_party/move/move-model/src/options.rs [hunk 15]
```diff
@@ -11,6 +12,9 @@ pub struct ModelBuilderOptions {
     /// part of the model.
```
Comment (ziaptos) [inline third_party/move/move-model/src/options.rs:12]: nit+: I get that this is not part of the PR - but this comment is confusing:
- What is considered "new compiler" - this will not be true forever.
- The variable name should include that info as well

File: third_party/move/move-prover/src/cli.rs [hunk 16]
```diff
@@ -67,7 +69,8 @@ pub struct Options {
     pub skip_attribute_checks: bool,
     /// Whether to use compiler v2 to compile Move code
     pub compiler_v2: bool,
-
+    /// The language version to use
+    pub language_version: Option<LanguageVersion>,
```
Comment (ziaptos) [inline third_party/move/move-prover/src/cli.rs:73]: nit: standardize with the rest of the code:
`maybe_language_version` or `optionjal_language_version`?

Also the comment suggest that the language version to use will be provided, what if its None?

File: aptos-move/framework/src/built_package.rs [hunk 17]
```diff
@@ -166,6 +177,25 @@ impl BuiltPackage {
     pub fn build(package_path: PathBuf, options: BuildOptions) -> anyhow::Result<Self> {
         let bytecode_version = options.bytecode_version;
         let compiler_version = options.compiler_version;
+        let language_version = options.language_version;
+        let effective_compiler_version = compiler_version.unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (wrwg) [reply aptos-move/framework/src/built_package.rs:181]: Need those `compiler_version` unwrapped instances a bit later in the code, hence can't overload the name. II moved this code into a helper function to avoid the issue.

File: third_party/move/move-model/src/metadata.rs [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct CompilationMetadata {
+    /// The version of the compiler, as a string.
+    pub compiler_version: String,
+    /// The version of the language, as a string.
+    pub language_version: String,
+    /// A flag indicating whether, at time of creation, the compilation
+    /// result was considered as unstable. Unstable code may have restrictions
+    /// for deployment on production networks.
+    pub unstable: bool,
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/metadata.rs:30]: Done

File: third_party/move/move-model/src/metadata.rs (hunk 11 above)
Comment (wrwg) [reply third_party/move/move-model/src/metadata.rs:217]: We can do this once we have more versions to deal with. It would be an easy local change.

File: third_party/move/move-model/src/model.rs (hunk 12 above)
Comment (wrwg) [reply third_party/move/move-model/src/model.rs:645]: We typically don't add the `is` qualifier.

File: third_party/move/move-model/src/model.rs [hunk 19]
```diff
@@ -637,11 +644,21 @@ impl GlobalEnv {
         self.generated_by_v2 = yes
     }
 
-    /// Returns if this is generated by v2.
+    /// Returns true if this is generated by v2.
     pub fn generated_by_v2(&self) -> bool {
         self.generated_by_v2
     }
 
+    /// Sets the language version.
```
[the remark this answers is not in the document]
Comment (wrwg) [reply third_party/move/move-model/src/model.rs:649]: Well, I used to do this 'redundant' comments anyway, to satisfy the principle that each public function has a comment. In my experience, if you don't do, people come along later and just add arbitrary undocumented functions, because they think its ok from other examples.

File: third_party/move/move-model/src/model.rs (hunk 13 above)
Comment (wrwg) [reply third_party/move/move-model/src/model.rs:654]: dito

File: third_party/move/move-model/src/options.rs (hunk 14 above)
Comment (wrwg) [reply third_party/move/move-model/src/options.rs:16]: Same here.

File: third_party/move/move-model/src/options.rs (hunk 15 above)
Comment (wrwg) [reply third_party/move/move-model/src/options.rs:12]: I removed mentioning of the compiler, its really what the variable says, that the move-model is used as the compiler. This flag will also go away once we retire the v1 compiler.

File: third_party/move/move-prover/src/cli.rs (hunk 16 above)
Comment (wrwg) [reply third_party/move/move-prover/src/cli.rs:73]: Actually all over the code (including Aptos CLI) we typically do not use `optional_...`. Changing this now would be inconsistent with 1/2 dozen other places which already existed before this PR. I personally think its OK to not encode this in the variable name.

**Comment by wrwg** [comment]
Thanks for the review!

**Review by vineethk — APPROVED** [review]
